googleapiclient/http.py

```diff
@@ -1200,9 +1205,10 @@ def _serialize_request(self, request):
     msg = MIMENonMultipart(major, minor)
     headers = request.headers.copy()
 
-    if request.http is not None and hasattr(request.http.request,
-        'credentials'):
-      request.http.request.credentials.apply(headers)
+    if request.http is not None:
+      credentials = _auth.get_credentials_from_http(request.http)
```

Comment:
There's a slight difference here compared to my PR.  It's been quite a while since I've looked at this, but the problem with GAE app-default credentials is that their `"access_token"` property is `None`.  This leads to problems when `credentials.apply(headers)` is called because the [`credentials.apply`](https://github.com/google/oauth2client/blob/master/oauth2client/client.py#L556) does a string concatenation without checking anything.

The actual proof that the credentials don't have an `access_token` is hard to trace through the code, but I'll try.

The app-default credentials are of type [`AppAssertionCredentials`](https://github.com/google/oauth2client/blob/9f0618d53481a3afe2d0cf232131ba21dcf583de/oauth2client/contrib/appengine.py#L119) which inherits from [`AssertionCredentials`](https://github.com/google/oauth2client/blob/master/oauth2client/client.py#L1444) which inherits from [`GoogleCredentials`](https://github.com/google/oauth2client/blob/master/oauth2client/client.py#L1048).  And that finally gets us to inheriting from `OAuth2Credentials` -- Which is the first concrete subclass of `Credentials` in this inheritance tree.  `GoogleCredentials` is documented to expect a string access token (first parameter in `__init__`), but if you look at the code of  `AppAssertionCredentials`, it [unconditionally passes `None`](https://github.com/google/oauth2client/blob/master/oauth2client/client.py#L1473).

In order to get this to actually work on GAE, you need to put something in your `_auth` wrapper to make sure that `apply_credentials` becomes robust to the case where the credentials have a `None` `access_token` value.  In my initial PR, I checked that the `credentials.access_token is not None` before calling `credentials.apply(headers)`.  In principle you could still do that in your auth wrapper after checking that the credentials are coming from the `oauth2client` library and not the `google_auth` library ...